### ISSUE TITLE
API: RandomState deprecation, default_rng, and SPEC7

### ISSUE THREAD

**Issue by larsoner** [mne-tools/mne-python#9233]
np.random.RandomState is not a good choice for random number generation [going forward](https://numpy.org/doc/stable/reference/random/legacy.html#numpy.random.RandomState):

> RandomState is effectively frozen and will only receive updates that are required by changes in the the internals of Numpy. More substantial changes, including algorithmic improvements, are reserved for Generator.

I don't understand all the details, but the TL;DR I've gathered from monitoring some SciPy discussions is that it's not as good as the new `default_rng` / `Generator`-based way of making random numbers and it should be replaced in people's code.

In the long term I think we want to switch `random_state=0` to use `default_rng(random_state)` instead of `RandomState(random_state)` internally. I propose:

1. Have a `mne.set_rng_backend(backend)` that can be `'RandomState'` or `'Generator'`, maybe with a `use_rng_backend` context manager. The context manager is maybe overkill, but I do think it might help people transition code from old to new, and it's only a few lines for us.
2. If this has not been called by the user yet, if they do `random_state=<int>` they will see a DeprecationWarning stating that the default is `'RandomState'` in 0.23 but will change to `'Generator'` in 0.24 (or maybe this is a good case for making it 0.25 to give people more time to change over), and we will call `set_rng_backend('RandomState')`.
3. In 0.24 (or 0.25) we no longer emit this warning and just use `'Generator'`.

I think this is a workable solution for modernizing our `random_state` params.

**Comment by drammock** [comment]
The new Generator class docstring says:

> No Compatibility Guarantee
> 
> Generator does not provide a version compatibility guarantee. In particular, as better algorithms evolve the bit stream may change.

I think it's worth asking what we use RNGs for.

- **reproducibility of examples / analyses**  Here, sticking with the frozen `RandomState` approach is actually *better* than switching to the new `Generator` / `default_rng` approach, right?
- **in our tests** This should be unaffected (usually for generating small fake data, the actual data values shouldn't matter)
- **permutations, bootstrapping, random parcellation**  Here it's not clear to me whether the old/new RNG approaches will make much difference to us, but my instinct is that the results will technically change (same script before/after will have slightly different results) but in practice it shouldn't actually matter.
- `numerics.random_permutation`  This is a utility designed to emulate MATLAB's `randperm` for sanity-check comparisions.  This will probably break?
- **`mxne`, ICA**  Here I don't think I understand well enough how the randomness is used to say what the situation is.

**Comment by larsoner** [comment]
I think for us we want everything to always be reproducible for a given int `random_state` (and a given `mne.set_rng_backend`). I hadn't looked at Generator / default_rng, it is indeed a problem that it does not guarantee stability.

I wonder if we should just switch to PCG64, which is guaranteed to be stable:

> Compatibility Guarantee
>
> PCG64 makes a guarantee that a fixed seed and will always produce the same random integer stream.

But is better than RandomState:

> By default, Generator uses bits provided by PCG64 which has better statistical properties than the legacy MT19937 used in RandomState.

**Comment by drammock** [comment]
agree on the plan to switch (with deprecation cycle) to explicitly use PCG64 (which happens be be the current default for `np.random.default_rng`, but that might change).

**Comment by agramfort** [comment]
scipy will deprecate RandomState ?

scikit-learn has not switched yet so I don't feel the urgence

>

**Comment by larsoner** [comment]
> scipy will deprecate RandomState ?

There is a long ongoing discussion in the SciPy PR [DOC: remove references to RandomState (#13778)](https://github.com/scipy/scipy/pull/13778). The short version so far to me is that RandomState will continue to work at least for some time but should not (and will not) be the preferred way to do things in examples and parameters.

**Comment by rkern** [comment]
Hi everyone! I'm the cause of everyone's problems here. As you work through these issues, I do recommend reading [NEP 19](https://numpy.org/neps/nep-0019-rng-policy.html) which goes into more detail about why we had the old backwards-compatibility policy and why we moved away from it. The concerns you have brought up here were definitely in our thoughts. I'm happy to answer any questions that the NEP doesn't answer.

I can't speak definitively as to when/how/if scipy really deprecates `RandomState` per se, but I have [my guess as to where we're going](https://github.com/scipy/scipy/pull/13778#issuecomment-811946326) on the aforementioned PR. Some time after we drop support for numpy 1.16, I expect that we will transition the code to use `Generator` instead of `RandomState`. I expect that we'll still _accept_ `RandomState` instances, but we will wrap its core uniform PRNG in `Generator` and call its methods instead. So if you call `scipy.stats.norm.rvs(random_state=RandomState(1234))`, you will get the results from `Generator.standard_normal()`'s improved algorithm rather than the older, slower `RandomState.standard_normal()`. The core Mersenne Twister PRNG underneath the passed `RandomState` will still be used, however. Again, we haven't made any decisions, yet, but I think that's where we are going.

**Comment by larsoner** [comment]
Thanks for the info @rkern . After some reading and discussion I think our plan is to shelve this for now until we bump our minimum numpy requirement to 1.17+ and then switch everything over to using Generator syntax internally (probably still using MT19937 for backward compat), which will likely happen in the next year.

**Comment by rkern** [comment]
Sounds like a good plan.

**Comment by larsoner** [comment]
If/when we do this we should probably also consider following [SPEC7](https://scientific-python.org/specs/spec-0007/) which I admittedly have not read yet but sounds relevant :slightly_smiling_face:

**Comment by sappelhoff** [comment]
Just a note that I have recently learned a bit more about setting RNG seeds thanks to this document: https://blog.scientific-python.org/numpy/numpy-rng/

And the current state in MNE-Python regarding this is, that we use a lot of RNG seeds that are **not recommended**. That is, seeds like 0, 1, 42, 97.

This is not only done in tests, but also in examples+tutorials, where I see this as a problem. We should set a good example in those scripts and use longer and more random seeds (not the common ones like 42, 1337, ...). @larsoner do you think we could/should touch those examples before tackling the brunt of this issue?

PS: When I recently tried to run my mne ICA with a seed generated by `secrets.randbits(128)`, I got an issue that the seed is too long. So that is also something that is not great and that we should fix.

**Comment by bruAristimunha** [comment]
I am interested if nobody assume

**Comment by bruAristimunha** [comment]
## RNG audit


## Shared RNG handling

| Source | Current behavior | Migration implication |
|---|---|---|
| [`check_random_state`](https://github.com/mne-tools/mne-python/blob/main/mne/utils/check.py#L212-L230) | `None` returns NumPy's global legacy `RandomState`; integers construct legacy `RandomState`; both `RandomState` and `Generator` instances are accepted. | Central transition point. |
| [`rng_uniform`](https://github.com/mne-tools/mne-python/blob/main/mne/fixes.py#L101-L104) | Bridges `Generator.integers` and `RandomState.randint`. | Remove when internal code uses `Generator` only. |
| [`random_state` docstring](https://github.com/mne-tools/mne-python/blob/main/mne/utils/docs.py#L3765-L3772) | Says `None` uses OS entropy. | Incorrect today: the implementation uses global state. |
| [`seed` docstring](https://github.com/mne-tools/mne-python/blob/main/mne/utils/docs.py#L4132-L4139) | Says `None` uses OS entropy. | Incorrect today for the same reason. |

## User-facing RNG parameters

| API group | Current argument | Source |
|---|---|---|
| Epoch equalization | `random_state` | [`BaseEpochs.equalize_event_counts`](https://github.com/mne-tools/mne-python/blob/main/mne/epochs.py#L2492-L2501), [`equalize_epoch_counts`](https://github.com/mne-tools/mne-python/blob/main/mne/epochs.py#L4015-L4020) |
| Epoch bootstrap | `random_state` | [`bootstrap`](https://github.com/mne-tools/mne-python/blob/main/mne/epochs.py#L4657-L4683) |
| Simulation | `random_state` | [`select_source_in_label`](https://github.com/mne-tools/mne-python/blob/main/mne/simulation/source.py#L24-L44), [`simulate_sparse_stc`](https://github.com/mne-tools/mne-python/blob/main/mne/simulation/source.py#L95-L130), [`simulate_evoked`](https://github.com/mne-tools/mne-python/blob/main/mne/simulation/evoked.py#L20-L55), [`add_noise`](https://github.com/mne-tools/mne-python/blob/main/mne/simulation/evoked.py#L102-L116), [`add_eog`](https://github.com/mne-tools/mne-python/blob/main/mne/simulation/raw.py#L391-L404), [`add_ecg`](https://github.com/mne-tools/mne-python/blob/main/mne/simulation/raw.py#L446-L459) |
| Labels | `random_state` | [`random_parcellation`](https://github.com/mne-tools/mne-python/blob/main/mne/label.py#L1953-L1974), [`select_sources`](https://github.com/mne-tools/mne-python/blob/main/mne/label.py#L2894-L2928) |
| ICA and inverse methods | `random_state` | [`ICA.__init__`](https://github.com/mne-tools/mne-python/blob/main/mne/preprocessing/ica.py#L430-L470), [`infomax`](https://github.com/mne-tools/mne-python/blob/main/mne/preprocessing/infomax_.py#L14-L82), [`mixed_norm`](https://github.com/mne-tools/mne-python/blob/main/mne/inverse_sparse/mxne_inverse.py#L345-L386) |
| Statistics | `seed` | [`permutation_t_test`](https://github.com/mne-tools/mne-python/blob/main/mne/stats/permutations.py#L25-L56), [`permutation_cluster_test`](https://github.com/mne-tools/mne-python/blob/main/mne/stats/cluster_level.py#L1106-L1155), [`permutation_cluster_1samp_test`](https://github.com/mne-tools/mne-python/blob/main/mne/stats/cluster_level.py#L1204-L1242), [`spatio_temporal_cluster_1samp_test`](https://github.com/mne-tools/mne-python/blob/main/mne/stats/cluster_level.py#L1314-L1355), [`spatio_temporal_cluster_test`](https://github.com/mne-tools/mne-python/blob/main/mne/stats/cluster_level.py#L1411-L1454) |
| Statistics and utility | `random_state` | [`bootstrap_confidence_interval`](https://github.com/mne-tools/mne-python/blob/main/mne/stats/permutations.py#L108-L148), [`random_permutation`](https://github.com/mne-tools/mne-python/blob/main/mne/utils/numerics.py#L269-L302) |

## Private and internal RNG use

| Area | Private implementation | RNG behavior | Migration note |
|---|---|---|---|
| Epoch equalization | [`_get_drop_indices`](https://github.com/mne-tools/mne-python/blob/main/mne/epochs.py#L4060-L4080) | `choice` after `check_random_state`. | Receive or create a `Generator`. |
| Simulated noise | [`_generate_noise`](https://github.com/mne-tools/mne-python/blob/main/mne/simulation/evoked.py#L171-L180) | Gaussian samples after `check_random_state`. | Receive a normalized `Generator`. |
| Simulated ECG/EOG | [`_add_exg`](https://github.com/mne-tools/mne-python/blob/main/mne/simulation/raw.py#L498-L579) | `uniform` and `normal` samples. | Receive a normalized `Generator`. |
| Label parcellation | [`_cortex_parcellation`](https://github.com/mne-tools/mne-python/blob/main/mne/label.py#L2003-L2032) | Random seed vertices via `choice`. | Receive a normalized `Generator`. |
| Infomax ICA | [`infomax`](https://github.com/mne-tools/mne-python/blob/main/mne/preprocessing/infomax_.py#L120-L228) | Random permutations and uniform samples. | Normalize once per fitting run. |
| MxNE | [`power_iteration_kron`](https://github.com/mne-tools/mne-python/blob/main/mne/inverse_sparse/mxne_debiasing.py#L13-L40), [`_compute_mxne_sure`](https://github.com/mne-tools/mne-python/blob/main/mne/inverse_sparse/mxne_inverse.py#L916-L1078) | Gaussian initialization/perturbation. | Migrate internal arguments directly. |
| Cluster permutation | [`_permutation_cluster_test`](https://github.com/mne-tools/mne-python/blob/main/mne/stats/cluster_level.py#L816-L980), [`_get_1samp_orders`](https://github.com/mne-tools/mne-python/blob/main/mne/stats/cluster_level.py#L774-L807) | Permutations, choices, and uniform sign flips. | Internal variable should consistently be `rng`. |
| Bootstrap CI | [`_ci`](https://github.com/mne-tools/mne-python/blob/main/mne/stats/permutations.py#L154-L161) | Forwards state for the bootstrap path. | Rename alongside public transition. |
| MATLAB compatibility utility | [`random_permutation`](https://github.com/mne-tools/mne-python/blob/main/mne/utils/numerics.py#L298-L302) | Uniform values plus `argsort`. | Preserve its explicitly MATLAB-compatible behavior in dedicated tests. |
| Circle-plot jitter | [`plot_connectivity_circle` internals](https://github.com/mne-tools/mne-python/blob/main/mne/viz/circle.py#L260-L276) | Direct fixed `RandomState(0)`. | Use a fixed local `Generator`; no public parameter is required. |
| Visualization sampling | [`_compute_scalings` internals](https://github.com/mne-tools/mne-python/blob/main/mne/viz/utils.py#L1398-L1411) | Fresh `default_rng()` selects display epochs. | Decide whether nondeterministic display-only behavior is acceptable. |

## Third-party RNG boundaries

| MNE-owned constructor/call | Source | Finding |
|---|---|---|
| sklearn PCA | [`_auto_low_rank_model`](https://github.com/mne-tools/mne-python/blob/main/mne/cov.py#L1506-L1522) | `PCA(random_state=None)` can choose randomized SVD and use sklearn/NumPy global state. This was verified by fitting identical data after different `np.random.seed` calls. |
| sklearn FactorAnalysis | [`_auto_low_rank_model`](https://github.com/mne-tools/mne-python/blob/main/mne/cov.py#L1514-L1521) | sklearn's default `random_state=0` is deterministic, but the seed policy is implicit. |
| sklearn LogisticRegression | [`LinearModel.fit`](https://github.com/mne-tools/mne-python/blob/main/mne/decoding/base.py#L537-L543) | `solver="liblinear"` is created without `random_state`; sklearn documents this state as controlling liblinear shuffling. |
| sklearn MultiTaskLasso | [`_mixed_norm_solver_cd`](https://github.com/mne-tools/mne-python/blob/main/mne/inverse_sparse/mxne_optim.py#L148-L161) | It has `random_state=None`, but MNE selects `selection="cyclic"`, so random coordinate selection is inactive. |
| sklearn FastICA / Picard / Infomax | [`ICA._fit`](https://github.com/mne-tools/mne-python/blob/main/mne/preprocessing/ica.py#L887-L981) | MNE correctly forwards its normalized state. |
